### PR TITLE
[ICD] Add ClientType support to the ICDManager and ICD Management cluster

### PR DESCRIPTION
**Pull request by mkardous-silabs**
#### Description
PR adds support for the ClientType in the ICDManager and the ICD Management cluster.

- Adds the ClientType to the RegisterClient command & the RegisteredClients attributes
- Adds the ClientType validation tothe ICDManager when checking for ICD Check-In messages
- Updates the ICDM 3.1 test plan to match test plan update
- Update to unit tests for the ClientType

Spec PR : 

- https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/8871
- https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/9315.
- https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/9322

Test Plan PR to adapt ICDM 3.1 test case : 

- https://github.com/CHIP-Specifications/chip-test-plans/pull/4200

#### Drive by changes :
PairingCommand used the wrong node-id for the ICDClientInfo storage. Modified the command sending to use to use the CheckInNodeID.

#### Note to reviewers
PR seems very large but isn't in reality. Since the XML was updates, the generated files updates is present in this PR as well.

#### Tests
- Unit tests
- Integration tests
- Manual tests to validate no Check-In message is sent out

### CODE REVIEW

File: src/app/clusters/icd-management-server/icd-management-server.cpp [hunk 1]
```diff
@@ -252,6 +253,7 @@ Status ICDManagementServer::RegisterClient(CommandHandler * commandObj, const Co
     FabricIndex fabricIndex            = commandObj->GetAccessingFabricIndex();
     NodeId nodeId                      = commandData.checkInNodeID;
     uint64_t monitoredSubject          = commandData.monitoredSubject;
+    ClientTypeEnum clientType          = commandData.clientType;
```
Comment (bzbarsky-apple) [inline src/app/clusters/icd-management-server/icd-management-server.cpp:256]: Something here needs to check that this is a valid ClientTypeEnum value, no?

File: src/app/icd/server/ICDCheckInSender.cpp [hunk 2]
```diff
@@ -27,10 +27,14 @@ namespace app {
 
 using namespace Protocols::SecureChannel;
 
-ICDCheckInSender::ICDCheckInSender(Messaging::ExchangeManager * exchangeManager)
+ICDCheckInSender::ICDCheckInSender(Messaging::ExchangeManager * exchangeManager, AddressResolve::Resolver * addressResolver)
```
Comment (bzbarsky-apple) [inline src/app/icd/server/ICDCheckInSender.cpp:30]: These changes for the address resolver seem unrelated to ClientType.  Shouldn't they be in a separate PR?

File: src/app/icd/server/ICDManager.cpp [hunk 3]
```diff
@@ -187,6 +192,13 @@ void ICDManager::SendCheckInMsgs()
                 continue;
             }
 
+            if (entry.clientType == ClientTypeEnum::kEphemeral)
+            {
+                // If the registered client is ephemeral, do not send a Check-In message
+                // continue to next entry
```
Comment (bzbarsky-apple) [inline src/app/icd/server/ICDManager.cpp:198]: I see no support for this behavior in the spec.  Spec says to send check-ins to all registered clients, regardless of type, no?

File: src/app/tests/suites/TestIcdManagementCluster.yaml [hunk 4]
```diff
@@ -208,6 +210,8 @@ tests:
                 value: 1001
               - name: "Key"
                 value: "\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f\xff"
+              - name: "ClientType"
+                value: 0
```
Comment (bzbarsky-apple) [inline src/app/tests/suites/TestIcdManagementCluster.yaml:214]: Please use symbolic enum names in the YAML.  Much more readable.  So `value: ClientType.Permanent` or `value: ClientType.Ephemeral`.

File: src/app/zap-templates/zcl/data-model/chip/icd-management-cluster.xml [hunk 5]
```diff
@@ -39,17 +39,24 @@ limitations under the License.
         <field name="AppDefinedButton" mask="0x10000"/>
     </bitmap>
 
+    <enum name="ClientTypeEnum" type="enum8">
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/icd-management-cluster.xml:42]: Spec does not say this is an enum8.  It says nothing about the sizing.  Needs to be fixed.

File: src/app/icd/server/ICDManager.cpp (hunk 3 above)
Comment (mkardous-silabs) [reply src/app/icd/server/ICDManager.cpp:198]: A sentence was added to say that an ICD should not send Check-In messages to Ephemeral clients.
https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/b3652909bdf595c7634cea1da0f1e58f837ea35f/src/data_model/ICDBehavior.adoc?plain=1#L223

File: src/app/zap-templates/zcl/data-model/chip/icd-management-cluster.xml (hunk 5 above)
Comment (mkardous-silabs) [reply src/app/zap-templates/zcl/data-model/chip/icd-management-cluster.xml:42]: Agreed, opened a spec issue for this last week 
https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/9281

File: src/app/tests/suites/TestIcdManagementCluster.yaml (hunk 4 above)
Comment (mkardous-silabs) [reply src/app/tests/suites/TestIcdManagementCluster.yaml]: done

File: src/app/icd/server/ICDCheckInSender.cpp (hunk 2 above)
Comment (mkardous-silabs) [reply src/app/icd/server/ICDCheckInSender.cpp]: done

File: examples/chip-tool/commands/clusters/ClusterCommand.h [hunk 6]
```diff
@@ -68,8 +69,9 @@ class ClusterCommand : public InteractionModelCommands, public ModelCommand, pub
                            const chip::app::Clusters::IcdManagement::Commands::RegisterClient::Type & value)
     {
         ReturnErrorOnFailure(InteractionModelCommands::SendCommand(device, endpointId, clusterId, commandId, value));
-        mScopedNodeId     = chip::ScopedNodeId(value.checkInNodeID, device->GetSecureSession().Value()->GetFabricIndex());
+        mScopedNodeId     = chip::ScopedNodeId(device->GetDeviceId(), device->GetSecureSession().Value()->GetFabricIndex());
```
Comment (yunhanw-google) [inline examples/chip-tool/commands/clusters/ClusterCommand.h:72]: we need to use checkin nodeId instead of peer nodeId, since we may register different nodeId for LIT.

File: examples/chip-tool/commands/clusters/ClusterCommand.h [hunk 7]
```diff
@@ -148,6 +150,7 @@ class ClusterCommand : public InteractionModelCommands, public ModelCommand, pub
                 clientInfo.peer_node         = mScopedNodeId;
                 clientInfo.monitored_subject = mMonitoredSubject;
                 clientInfo.start_icd_counter = value.ICDCounter;
+                clientInfo.client_type       = mClientType;
```
Comment (yunhanw-google) [inline examples/chip-tool/commands/clusters/ClusterCommand.h:153]: We should not store client type in icd clientInfo, since application knows the client type, and we don't receive any check-in message for ephemeral, i cannot figure out when and how to use clientInfo for which purpose..

**Review by yunhanw-google — CHANGES_REQUESTED** [review]

File: src/app/icd/server/ICDManager.cpp (hunk 3 above)
Comment (yunhanw-google) [reply src/app/icd/server/ICDManager.cpp:194]: I think we should still allow to send check-in message when manually wake-up device so that controller, particularly mobile controller, has the way to know the LIT becomes active, and trigger the customized strategy if controller implements it, for example, send several stayActive command for x seconds, for example, user uses mobile controller to interact with one sensor in tough place, they may wanna interact with this lit for longer period instead of manually wake-up back and forth..

Comment (yunhanw-google) [reply src/app/icd/server/ICDManager.cpp:194]: @bzbarsky-apple @mkardous-silabs @turon

File: examples/chip-tool/commands/clusters/ClusterCommand.h (hunk 6 above)
Comment (mkardous-silabs) [reply examples/chip-tool/commands/clusters/ClusterCommand.h:72]: In that case, what is stored during the commissioning registration is incorrect - the ICDs NodeId was being stored.

I'll revert this changes and fix the PairingCommand bits.

File: examples/chip-tool/commands/clusters/ClusterCommand.h (hunk 7 above)
Comment (mkardous-silabs) [reply examples/chip-tool/commands/clusters/ClusterCommand.h:153]: I am not sure i agree with the statement.
Not persisting the ClientType forces an all or nothing scenario. A client will be only be able to use one type. If it reboots, the only way it knows the type it used at registration if it is not persisted is by always using the same type.

The registration ClientType affects the assumptions a client can do ie. is it garanteed a check-in message will be sent or not and as such we should make sure this information is always available.

> i cannot figure out when and how to use clientInfo for which purpose.

If you do not believe it is necessary, nothing prevents you from not using the information but we shouldn't assume that others might not find a benefit of having this information persisted. It is only a uint8 so the cost is pretty minimal and can enable several use-case we might have missed.

File: src/app/icd/server/ICDManager.cpp (hunk 3 above)
Comment (mkardous-silabs) [reply src/app/icd/server/ICDManager.cpp:194]: How do you differentiate between a user using his device for normal operations vs. triggering the user active mode trigger do interact from the mobile only controller?

File: examples/chip-tool/commands/clusters/ClusterCommand.h (hunk 6 above)
Comment (yunhanw-google) [reply examples/chip-tool/commands/clusters/ClusterCommand.h:72]: yes, please fix commissioning registration and allow to set the other nodeId, thanks

Comment (mkardous-silabs) [reply examples/chip-tool/commands/clusters/ClusterCommand.h]: done

File: src/app/zap-templates/zcl/data-model/chip/icd-management-cluster.xml (hunk 5 above)
Comment (mkardous-silabs) [reply src/app/zap-templates/zcl/data-model/chip/icd-management-cluster.xml:42]: https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/9322

File: src/app/clusters/icd-management-server/icd-management-server.cpp (hunk 1 above)
Comment (mkardous-silabs) [reply src/app/clusters/icd-management-server/icd-management-server.cpp:256]: done

File: src/app/icd/client/DefaultICDClientStorage.cpp [hunk 8]
```diff
@@ -274,6 +274,10 @@ CHIP_ERROR DefaultICDClientStorage::Load(FabricIndex fabricIndex, std::vector<IC
         memcpy(clientInfo.hmac_key_handle.AsMutable<Crypto::Symmetric128BitsKeyByteArray>(), hmacBuf.data(),
                sizeof(Crypto::Symmetric128BitsKeyByteArray));
 
+        // ClientType
+        ReturnErrorOnFailure(reader.Next(TLV::ContextTag(ClientInfoTag::kClientType)));
+        ReturnErrorOnFailure(reader.Get(clientInfo.client_type));
```
[the remark this answers is not in the document]
Comment (mkardous-silabs) [reply src/app/icd/client/DefaultICDClientStorage.cpp:279]: Since the feauture is still provisionnal, i would say yes. Once LIT is officially out, it won't be.

File: examples/chip-tool/commands/clusters/ClusterCommand.h (hunk 7 above)
Comment (yunhanw-google) [reply examples/chip-tool/commands/clusters/ClusterCommand.h:153]: Ack, thanks

File: src/app/icd/client/DefaultICDClientStorage.h [hunk 9]
```diff
@@ -130,6 +130,7 @@ class DefaultICDClientStorage : public ICDClientStorage
         kMonitoredSubject = 5,
         kAesKeyHandle     = 6,
         kHmacKeyHandle    = 7,
+        kClientType       = 8,
```
Comment (yunhanw-google) [inline src/app/icd/client/DefaultICDClientStorage.h:133]: please also update MaxICDClientInfoSize in https://github.com/project-chip/connectedhomeip/blob/master/src/app/icd/client/DefaultICDClientStorage.h#L159

File: src/app/clusters/icd-management-server/icd-management-server.cpp [hunk 10]
```diff
@@ -252,10 +253,14 @@ Status ICDManagementServer::RegisterClient(CommandHandler * commandObj, const Co
...
     ByteSpan key                       = commandData.key;
     Optional<ByteSpan> verificationKey = commandData.verificationKey;
     bool isClientAdmin                 = false;
 
+    // Check if ClientType is valid
+    VerifyOrReturnError(clientType < ClientTypeEnum::kUnknownEnumValue, InteractionModel::Status::ConstraintError);
```
[the remark this answers is not in the document]
Comment (mkardous-silabs) [reply src/app/clusters/icd-management-server/icd-management-server.cpp]: Done - added a negative test also for the `kUnknownEnumValue`

File: examples/chip-tool/commands/pairing/PairingCommand.h [hunk 11]
```diff
@@ -71,6 +71,8 @@ class PairingCommand : public CHIPCommand,
                     "The check-in node id for the ICD, default: node id of the commissioner.");
         AddArgument("icd-monitored-subject", 0, UINT64_MAX, &mICDMonitoredSubject,
                     "The monitored subject of the ICD, default: The node id used for icd-check-in-nodeid.");
+        AddArgument("icd-client-type", 0, 1, &mICDClientType,
+                    "The ClientType of the client regsitering, default: Permanent client - 0");
```
[the remark this answers is not in the document]
Comment (mkardous-silabs) [reply examples/chip-tool/commands/pairing/PairingCommand.h]: done

File: examples/fabric-admin/commands/pairing/PairingCommand.h [hunk 12]
```diff
@@ -70,6 +70,8 @@ class PairingCommand : public CHIPCommand,
                     "The check-in node id for the ICD, default: node id of the commissioner.");
         AddArgument("icd-monitored-subject", 0, UINT64_MAX, &mICDMonitoredSubject,
                     "The monitored subject of the ICD, default: The node id used for icd-check-in-nodeid.");
+        AddArgument("icd-client-type", 0, 1, &mICDClientType,
+                    "The ClientType of the client regsitering, default: Permanent client - 0");
```
[the remark this answers is not in the document]
Comment (mkardous-silabs) [reply examples/fabric-admin/commands/pairing/PairingCommand.h]: done

File: src/app/icd/client/DefaultICDClientStorage.h (hunk 9 above)
Comment (mkardous-silabs) [reply src/app/icd/client/DefaultICDClientStorage.h:133]: done

File: src/app/tests/suites/TestIcdManagementCluster.yaml [hunk 13]
```diff
@@ -221,6 +225,23 @@ tests:
                 value: 1001
               - name: "Key"
                 value: "\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x1e\x1f"
+              - name: "ClientType"
+                value: ClientTypeEnum.UnknownEnumValue
```
Comment (bzbarsky-apple) [inline src/app/tests/suites/TestIcdManagementCluster.yaml:229]: You could also consider sending some other out-of-range value if the YAML will let you (like 3 or something).  That should also result in ConstraintError.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by yunhanw-google — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

File: src/app/tests/suites/TestIcdManagementCluster.yaml (hunk 13 above)
Comment (mkardous-silabs) [reply src/app/tests/suites/TestIcdManagementCluster.yaml:229]: done

Comment (mkardous-silabs) [reply src/app/tests/suites/TestIcdManagementCluster.yaml:229]: @bzbarsky-apple When adding the test case with a value > UnknownEnumValue, yaml parsing fails since it checks the valid values of the enum. Couldn't add the test case with the yaml test.

This will need to go in the python test - will do in a follow up.
